### ISSUE TITLE
Intermittent failures in `mobilenetv3_fake_weights.mlir`

### ISSUE THREAD

**Issue by kuhar** [iree-org/iree#12444]
This test started failing intermittently after the most recent integrate (#12419).

Sample failure: https://github.com/openxla/iree/actions/runs/4307115731/jobs/7511785781#step:4:8565

Error:
```
FAILED: tests/e2e/models/check_cuda_cuda_mobilenetv3_fake_weights.mlir_module.vmfb /work/full-build-dir/tests/e2e/models/check_cuda_cuda_mobilenetv3_fake_weights.mlir_module.vmfb 
cd /work/full-build-dir/tests/e2e/models && /work/full-build-dir/tools/iree-compile --output-format=vm-bytecode --mlir-print-op-on-diagnostic=false --iree-hal-target-backends=cuda --iree-input-type=mhlo /work/tests/e2e/models/mobilenetv3_fake_weights.mlir -o check_cuda_cuda_mobilenetv3_fake_weights.mlir_module.vmfb --iree-hal-executable-object-search-path=\"/work/full-build-dir\" --iree-llvm-embedded-linker-path=\"/work/full-build-dir/llvm-project/bin/lld\" --iree-llvm-wasm-linker-path=\"/work/full-build-dir/llvm-project/bin/lld\"
/work/tests/e2e/models/mobilenetv3_fake_weights.mlir:933:12: error: cannot replace an op with another op producing a different number of results while tracking handles
    %695 = "mhlo.reduce"(%694, %266) ( {
           ^
/work/tests/e2e/models/mobilenetv3_fake_weights.mlir:938:12: error: 'hal.executable.variant' op Listener failed.
    %696 = mhlo.divide %695, %260 : tensor<1x288xf32>
           ^
/work/tests/e2e/models/mobilenetv3_fake_weights.mlir:938:12: error: Failed to performCanonicalization after transform
    %696 = mhlo.divide %695, %260 : tensor<1x288xf32>
           ^
/work/tests/e2e/models/mobilenetv3_fake_weights.mlir:938:12: error: transform dialect interpreter failed
    %696 = mhlo.divide %695, %260 : tensor<1x288xf32>
           ^
/work/tests/e2e/models/mobilenetv3_fake_weights.mlir:938:12: error: failed to run translation of source executable to target executable for backend #hal.executable.target<"cuda", "cuda-nvptx-fb", {target_arch = "sm_35"}>
    %696 = mhlo.divide %695, %260 : tensor<1x288xf32>
           ^
/work/tests/e2e/models/mobilenetv3_fake_weights.mlir:938:12: error: failed to serialize executables
    %696 = mhlo.divide %695, %260 : tensor<1x288xf32>
           ^
```

**Comment by kuhar** [comment]
From @matthias-springer on Discord:

---
Matthias — Today at 2:55 PM
I've seen a similar error recently when working on #12358.  These started showing up because of additional notifications. Notifications such as "op created", "op removed" and "op modified". There were a multiple cases in the GreedyPatternRewriteDriver and RewriterBase where notifications were not sent when they should have been.

The fact that these are causing issues now (in particular in some non-deterministic way) indicate that:
* We may have handles to deleted operations in the transform dialect interpreter. (dangling pointer) Or something else that causes undefined behavior.
* It does not cause any other crash, so these handles are not used.

My best guess that there is either
a) some RewritePatterns that we use as part of transform.apply_patterns etc. do not use the RewriterBase API correctly and directly modify IR.
b) RewriterBase API is used correctly but the implementation of RewriterBase does not handle notifications correctly. (E.g., RewriterBase::eraseOp does currently not send "op erased" notifications for nested ops.) I have a few fixes out for review on Phabricator for these issues, but there are still more that are not fixed yet.
The proper fix (if my guess is correct) is to fix the RewriterBase API. In particular, RewriterBase::eraseOp must be rewritten. Some of these fixes are non-trivial and since this is touching core MLIR I expect these to take a few weeks to go through review.
What we can do short-term is to drop op handle mappings from the transform interpreter dialect when there are no more future uses of the handle. I think think this would fix the concrete issue here.
Matthias — Today at 3:02 PM
I talked about this with Alex a few days ago and he mentioned that we had this functionality in the transform dialect interpreter some while ago. But he removed it because it can hide bugs. Bugs like the one here, which are most likely not issues with the transform dialect but the RewriterBase implementation in MLIR.

**Comment by nicolasvasilache** [comment]
Here is a list of ongoing mitigations:
  1. @matthias-springer has a draft PR to help detect patterns that do not properly use the rewriter API at runtime [link](https://reviews.llvm.org/D144552).
  2. we also sync'ed with @matthias-springer today and I suggested a bigger hammer: make all IR modification APIs on Operation/Value/Block private and expose a very few select friends to force everyone to funnel through `OpBuilder`/`RewriterBase`. This should turn all bad usages into compiler failures and I expect will also let us clean up a lot of things on the IREE side. I don't imagine this can land upstream given past experience with such proposed change but I believe as a draft PR to get a static version of pont 1. , it makes sense.
  3. on the pure transform-dialect side, we had a cleanup procedure that deleted the tracking of handles that went out of scope, this was retired not too long ago but we could think of reactivating to provide an additional safety net (@ftynse).
  4. on our usage, I have some ongoing cleanups to remove blanket canonicalizations and hoisting patterns now that we can control things more finely. I am prioritizing those on my end to reduce the amount of such manipulations that happen automatically under the hood (#12465 and #12467).
 
We'll post links as available during the investigation.

**Comment by ftynse** [comment]
> I talked about this with Alex a few days ago and he mentioned that we had this functionality in the transform dialect interpreter some while ago. But he removed it because it can hide bugs. Bugs like the one here, which are most likely not issues with the transform dialect but the RewriterBase implementation in MLIR.

> on the pure transform-dialect side, we had a cleanup procedure that deleted the tracking of handles that went out of scope, this was retired not too long ago but we could think of reactivating to provide an additional safety net

The removal happend a while ago https://github.com/llvm/llvm-project/commit/6bb997c032f509d4f61461aa461c8431325cbb2a. I don't mind temporarily putting it back as a hotfix, if necessary. It was very surprising for debugging, but now we are better with debug messages so a simple "not assigning because there are no further use" could do the trick. Longer term, the assumption that there is a single block that we can just traverse to see if there are further uses is untenable. The proper way to do it would be to do some sort of reaching definitions analysis and unmap handles that are no longer reachable on any of the possible control flow paths.

We _are_ removing handles that go out of lexical scope, https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Transform/IR/TransformInterfaces.h#L206. But this happens too late for what you may want here.

**Comment by ftynse** [comment]
The intermittency of the problem may be due to vicious pointer reuse, I've seen a couple of cases as follows: we have a handle to `Operation *`, the operation gets erased/replaced without us updating the handle, and another operation gets the same pointer address. Depending on the kind of the new operation, we may even "successfully" detect the "replacement".

**Comment by GMNGeoffrey** [comment]
@aaron-schneider said you were trying to figure out how to run the test many times to catch the flake. Rather than running the entire CI multiple times, which is very wasteful. Run the specific *test* multiple times. ctest has the flag `--repeat-until-fail`. Is this only reproducible on the CI? If not, the easiest thing is to just run ctest with that test and use that flag. If it is only happening on CI, you can hack the CI to do the same thing. We already run some tests multiple times as part of the TSAN build here: https://cs.github.com/openxla/iree/blob/5740497cc68cb3bef22d9ec8092f3e10009457b9/build_tools/cmake/ctest_all.sh?q=IREE_CTEST_REPEAT_UNTIL_FAIL_COUNT. We don't really want to run all the tests multiple times though, so you can edit the build script to only run that test.

**Comment by ScottTodd** [comment]
The list of "list of ongoing mitigations" posted above is great, thank you.

We need to increase test coverage here though (rule of thumb: every p0 bug deserves tests that at least would have caught that specific bug). Re-running a single suspected-flaky test is not enough - we need to understand the problem well enough to catch such bugs reliably.

It's concerning that only one full program test exhibited this buggy behavior, when we have a larger suite of full programs that we run benchmarks on and a much larger suite of unit tests. Something isn't adding up.

**Comment by GMNGeoffrey** [comment]
Oh this isn't actually a test, it's a build. That makes running multiple times less trivial. I think easiest would be to hack the build script to run build in a loop, deleting the output file each time.

But also Scott is right. The fact that we don't understand what the failure was well enough to add a test for it is in itself troubling

**Comment by aaron-schneider** [comment]
https://github.com/openxla/iree/actions/runs/4318850317 - the CI failed, but none of the individual jobs failed? Does that mean we should try rerunning? 

```
execution_benchmarks / run_benchmarks (c2-standard-16, linux, x86_64)GitHub Actions has encountered an internal error when running your job.
--
summaryGitHub Actions has encountered an internal error when running your job.

[execution_benchmarks / run_benchmarks (c2-standard-16, linux, x86_64)](https://github.com/openxla/iree/actions/runs/4318850317/jobs/7538122537)
GitHub Actions has encountered an internal error when running your job.
[summary](https://github.com/openxla/iree/actions/runs/4318850317/jobs/7538272394)
GitHub Actions has encountered an internal error when running your job.
```

**Comment by GMNGeoffrey** [comment]
GitHub actions are having issues apparently. See https://githubstatus.com

**Comment by GMNGeoffrey** [comment]
You can rerun only the failed job and see if it recovers

**Comment by MaheshRavishankar** [comment]
Just stepping back a bit, it looks like there are some issues related transform dialect that need to be addressed.

@nicolasvasilache @ftynse could you guys help give an estimate of what the timeline is here. This came about on integrate, but it doesn't look like the integrate was the root cause, rather it just uncovered a bug? The failure led to a revert of the integrate since it failed on one of the core tests on CUDA for the part that used the transform dialect lowering. If we want to reliably avoid such regressions there is only one option, and that is to disable the use of transform dialect for the CUDA path that is enabled by default now. We reverted the integrate so that we don't go down that path since that usage had graduated to a core IREE feature, but it does seem like we might have to till the issues uncovered here are addressed? What do you guys think? Going back here might lead to regressions downstream, so would prefer to not disable transform dialect usage but stabilitt/timeline will help make decision.

**Comment by matthias-springer** [comment]
This issue started to appear after my modifications to the upstream Listener/notifications infrastructure. In particular, more notifications are getting sent. There are a few more changes that have not gone through reviews yet and also potentially broken rewrite patterns that can (generally speaking) cause such issues.

I think this particular issue here was fixed by #12464. I cannot be entirely sure because the issue was not reproducible even after many runs.

My suggestion would be:
* Testing all notification/listener related changes with IREE before landing. (It would probably not have prevented this issue though.)
* Building additional debug/sanitizer infrastructure to detect such non-deterministic failures. E.g., [D144552](https://reviews.llvm.org/D144552). We can probably also add additional "expensive-checks" to the `transform` dialect that we run every once in a while.
* Long term: Adopt Mehdi's idea of [IR listeners](https://discourse.llvm.org/t/rfc-introduce-the-concept-of-ir-listeners-in-mlir/67854). It will not only make the transform dialect usage more robust but also guard against broken patterns such as the ones in #12466.

**Comment by ftynse** [comment]
> Is this only reproducible on the CI? 

@GMNGeoffrey FWIW, I tried running the equivalent command (`iree-compile --output-format=vm-bytecode --mlir-print-op-on-diagnostic=false --iree-hal-target-backends=cuda --iree-input-type=mhlo tests/e2e/models/mobilenetv3_fake_weights.mlir  -o check_cuda_cuda_mobilenetv3_fake_weights.mlir_module.vmfb --iree-hal-executable-object-search-path=$HOME/iree/build --iree-llvm-embedded-linker-path=HOME/iree/build/llvm-project/bin/lld`) machines, and none of them failed. Maybe there is something specific in the CI environment that makes it fail. Any chance of logging into the CI machine to test that without interfering with the CI operations?

> We need to increase test coverage here though.

Yep. The e2e test runs the tested compiler as part of the build process. I'm surprised there is no equivalent test that would just verify the output of the tested compiler. Or maybe there is, but it never gets to run. In any case, it sounds like a two-stage testing process would be preferable.

One random thought by starting at my repro attempts, I've seen `iree-compile` request 20T of virtual memory. Can it be running out-of-memory in the test environment? Or running close to that and therefore forcing more aggressive address reuse?

**Comment by ftynse** [comment]
> Just stepping back a bit, it looks like there are some issues related transform dialect that need to be addressed.

I would claim that the issues have very little to do with the transform dialect. At most, with the blanket auto-canonicalization that a specific transform op in LinalgExt is using. I repeatedly argued to make this less obnoxious, and this is finally happening in https://github.com/openxla/iree/pull/12467.

Since I haven't been able to reproduce this yet, it's hard to be certain, but the root cause is likely in malformed canonicalization patterns that bypass the rewriter infrastructure as @matthias-springer points out. This would remain a problem even if you disable the transform dialect compilation pipeline, it would be just hidden and appear again when upstream tweaks, e.g., the order in which canonicalizations are applied. 

> @nicolasvasilache @ftynse could you guys help give an estimate of what the timeline is here. 

@MaheshRavishankar it is unclear the estimate of what exactly you'd like to hear.

There is exactly one code path that I can see leading to the specific error being emitted. If you just want the integration to be unblocked, this can be done with a variety of hotfixes in O(hours).

Landing https://github.com/openxla/iree/pull/12467 after fixing the tests it breaks is O(days).

Fixing the canonicalization patterns, including the upstream ones, is O(weeks).

I have no idea what the project wants to do with this, I would have gone with the fastest fix to unblock progress and worked on a cleaner fix later. It seems that people more involved in IREE day-to-day lean towards one of the two other options.

**Comment by ftynse** [comment]
The hotfix in question is along the lines of

```diff
diff --git a/llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/IR/StructuredTransformOpsExt.cpp b/llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/IR/StructuredTransformOpsExt.cpp
index cf3cd226e..1c8f0a81b 100644
--- a/llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/IR/StructuredTransformOpsExt.cpp
+++ b/llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/IR/StructuredTransformOpsExt.cpp
@@ -452,6 +452,11 @@ void mlir::TrackingListener::notifyOperationReplaced(Operation *op,
     return;
   }
 
+  // If this would cause an error with replacement, drop instead.
+  if (*replacement && (*replacement)->getNumResults() != op->getNumResults()) {
+    replacement = nullptr;
+  }
+
   if (*replacement == nullptr) {
     // TODO: Check if the handle is dead. Otherwise, the op should not be
     // dropped. This needs a change in the transform dialect interpreter.
```

This is the only place where `replacePayloadOp` is called with the non-null replacement, which in turn is the only code path that emits the error message seen in the log https://github.com/llvm/llvm-project/blame/7d9be4368a36aab44f2f3b2afe8262ce129851e1/mlir/lib/Dialect/Transform/IR/TransformInterfaces.cpp#L345-L346.

**Comment by nicolasvasilache** [comment]
#12492 is #12465 rebase on the integrate #12475
Wasn't able to find errors even with hard asserts that we recover form (I deactivated those in my last commit).

As far as I am concerned, this is good to go.

Proof of non-existence can only be achieved with compiler support IMO: either the IR type system is listened to (e.g. w/ Mehdi's proposal) or the IR manipulation APIs are made private and forced to access through the RewriterBase. In the absence of that, bugs can be fixed as they surface.

Let's not fool ourselves though, in the realm of compiler bugs and problematic abstraction layerings that are bound to surface in the future, this is quite mild and appeared because we went for a quite deep stress test of the system: it is a good thing to surface these bugs and fix them.

Still, the last commit I added reverts the hard failures on errors for the purpose of not crashing when we can recover.
This has a similar flavor to Matthias' https://reviews.llvm.org/D144552.

I would prefer to turn it on by default and report future errors aggressively but this will show up in the CI and apparently that is a problem ..

How about we have a special IREE flag + CI mode that:
1. would be non-blocking 
2. trigger expensive/intrusive asserts from which we can recover in the "production" flow
3. still yield valuable information asynchronously to automatically find broken patterns that we can iterate on fixing?

I believe such a strict CI mode would be valuable and provide strong incentives for good hygiene in existing and future patterns.

**Comment by ScottTodd** [comment]
> * Testing all notification/listener related changes with IREE before landing. (It would probably not have prevented this issue though.)

If upstream changes can't be made confidently then upstream is missing test coverage too. Are there integration tests upstream? (See "two unit tests, no integration tests": [meme 1](https://twitter.com/ThePracticalDev/status/687672086152753152), [meme 2](https://twitter.com/timbray/status/822470746773409794))

+1 to the ideas about stricter checks and sanitizer modes (and fuzzing would be good too).

I'd like to understand why `mobilenetv3_fake_weights` was the only test where we noticed failures. We have multiple large frontends capable of importing complex programs (TFLite/TOSA, TF/MHLO, PyTorch/Linalg, JAX/MHLO, StableHLO, etc.). We should have more tests in-tree that at least capture the structure and complexity of programs produced by those frontends (https://github.com/openxla/iree/pull/12017 is a good start for JAX, but we need more).

I'd like to see more test cases like this one: https://github.com/openxla/iree/blob/9813b2af39c1079cb49ae194291856b180a33be7/runtime/src/iree/vm/bytecode/module_test.mlir#L20-L27 Whenever we see a frontend produce something silly / at a boundary condition, we should capture that in a test to ensure the compiler can safely handle it (and ideally handle it efficiently, but correctness is essential).

**Comment by MaheshRavishankar** [comment]
Oh geez! I tried very hard to word my comments to not ruffle feathers about transform dialect. I did not intend to throw mud on any of that work (clearly I failed). I am not making a distinction on whether a feature is worth it or not. I am only trying to make a distinction between a stable feature (default on) and an unstable feature (default off). The unstable feature might be hitting latent issues unrelated to the feature itself, but still is unstable. All software has bugs, newer code has potentially more bugs than older code.
So all I was trying to do is ascertain if a feature that was expected to be stable is there yet or not (for whatever reason). If there are intermittent errors that happen on unrelated changes then is there a way around marking it as unstable and not have unrelated changes hit hard to triage bugs? (Granted, even existing code could hit new intermittent failures, but those are harder to "turn off" and maybe be better to fix forward.)

> would prefer to turn it on by default and report future errors aggressively but this will show up in the CI and apparently that is a problem ..
> 
> How about we have a special IREE flag + CI mode that:
> 
> 1. would be non-blocking
> 2. trigger expensive/intrusive asserts from which we can recover in the "production" flow
> 3. still yield valuable information asynchronously to automatically find broken patterns that we can iterate on fixing?

This could only work for tests that are post-submit so that it doesn't block unrelated changes but gives a signal. I am not sure we have the CI resources to do this though.... 

Looking through responses above, this might all be moot. We can keep status quo and see if we hit similar failures later on. If we hit such a failure again then we could turn of the feature by default and triage more.

**Comment by GMNGeoffrey** [comment]
> I tried running the equivalent command [on multiple?] machines, and none of them failed.

Did you run inside the docker container used for CI?

**Comment by ftynse** [comment]
Thanks for clarifications, @MaheshRavishankar!

I am not sure about your definition of stability. It sounds very much like "not allowed to change, ever", at which point none of MLIR can be deemed stable. I suppose it is not what you really meant so I would rather expand from your last point: something can be seen as more stable if the rate at which it causes issues is low. More data points would be needed to estimate that rate.

If you are interested in the possible/planned changes to the transform dialect infrastructure, the core interpreter part is not expected to change much at this point. There will be additional features on matchers and runtime dispatches in the following months, but they should be minimally intrusive. I may expect changes around vectorization and packing operations to make them more flexible. There is always the unknown factor of a third-party contributor that we can't predict. IMO, it doesn't make the code any less stable than, say, the LLVM translation that has been undergoing changes in the past month.

Also, ironically, it is the older code that may be problematic here: some patterns predate changes in `PatternRewriter` and are using lower-level APIs.

**Comment by MaheshRavishankar** [comment]
> Thanks for clarifications, @MaheshRavishankar!
> 
> I am not sure about your definition of stability. It sounds very much like "not allowed to change, ever", at which point none of MLIR can be deemed stable. I suppose it is not what you really meant so I would rather expand from your last point: something can be seen as more stable if the rate at which it causes issues is low. More data points would be needed to estimate that rate.
> 

That cannot be the interpretation of stable. Cant change ever is not what I meant. Newer code is bound to have bugs (mostly w.r.t to how it interacts with existing things for this case). I would define stable as 
- has been tested on CI for a while
- It does not lead to failures on unrelated changes (even if the failure is w.r.t how the old code interacts with new code).
- Any such failures are easy to fix and are fixed forward
In this case it comes down to can you isolate regressions to a particular thing that can be turned off without **too** much downstream effects. Things that are around for a while potentially have more dependencies, and need to be fixed forward. If it is something that we can isolate, turn off, fix, turn on without too much issues downstream, that seems practical to me.

> If you are interested in the possible/planned changes to the transform dialect infrastructure, the core interpreter part is not expected to change much at this point. There will be additional features on matchers and runtime dispatches in the following months, but they should be minimally intrusive. I may expect changes around vectorization and packing operations to make them more flexible. There is always the unknown factor of a third-party contributor that we can't predict. IMO, it doesn't make the code any less stable than, say, the LLVM translation that has been undergoing changes in the past month.
> 
> Also, ironically, it is the older code that may be problematic here: some patterns predate changes in `PatternRewriter` and are using lower-level APIs.

I completely agree with this! Like I said this is exposing latent bugs, but given that we can isolate it to one thing with lesser dependencies, we have the option to turn off, fix, turn on, rather than having downstream users fail intermittently and adding huge pressure to fix forward...

**Comment by allieculp** [comment]
@nicolasvasilache @ScottTodd @GMNGeoffrey @ftynse @MaheshRavishankar Do we need to leave this open? Curious if there is more work to do here.

**Comment by ScottTodd** [comment]
* The specific bug here was fixed
* Some safeguards were put in place upstream, but similar bugs are still possible
* We're still a bit thin on test coverage in IREE (it's concerning that only one test was failing, it would have been nicer if we had a more diverse set of tests exercising those code paths)
